### ISSUE TITLE
Type assertion on the whole action

### ISSUE THREAD

**Issue by farzadmf** [piotrwitek/react-redux-typescript-guide#61]
Hi,

I read about the [typesafe-actions](https://github.com/piotrwitek/typesafe-actions) here, and I asked the question in there, but thought maybe it wouldn't hurt if I ask it here too.

I think the library it is what I have been looking for to substantially reduce boilerplate code for actions and action creators; thank you!

Creating actions (using `createAction`) and `switch`ing on them (using `getType`) in the reducer works fine, but I was wondering if there's a way to do type checking for a method that expects a specific kind of action.

Example of my actions (I'm using lodash for `values` and `map`):

```ts
export enum ActionTypes {
  GetCameras = 'GET_CAMERAS',
  GetCamerasSuccss = 'GET_CAMERAS_SUCCESS',
}

export const cameraActions = {
  getCameras: createAction(ActionType.GetCameras, ( /*...*/ ) => ({
    type: ActionTypes.GetCameras,
    /* payload: ... */
  }),
  getCamerasSuccess: createActionTypes.GetCamerasSuccess, (/*...*/) => ({
    type: ActionTypes.GetCameras,
    /* payload: ... */
  }),
};

const returnsOfActions = _.map(_.values(actions), $call));
export type CameraAction = typeof returnsOfActions[number];
```

Maybe a good example would be a case where the reducer wants to call another method for a specific action type, for example, in my reducer:

```ts
export const reducer = (state = initialState, action: CameraAction) => {
  case getType(cameraActions.getCameras):
    return state; /* ... */
  case getType(cameraActions.getCamerasSuccess):
    return receiveCameras(state, action);
  default:
    return state;
};
```

Now, in my `receiveCameras` method, I want to make sure the second argument is of type `cameraActions.getCamerasSuccess`.

I can do something like this:

```ts
const receiveCameras = (state: CameraState, action: CameraAction) => {
  if (action.type !== getType(cameraActions.getCamerasSuccess)) {
    return;
  }
  
  // Now my action type is correct
}
```

I was wondering if I can specify the type when defining the parameter in the method to avoid doing that `if` check

Thank you

**Comment by hoeck** [comment]
Looks like the next Typescript version will provide a solution for this https://blogs.msdn.microsoft.com/typescript/2018/03/15/announcing-typescript-2-8-rc/ via _conditional types_ and `ReturnType` in `lib.d.ts`:

```
const receiveCameras = (
    state: CameraState,
    action: ReturnType<typeof cameraActions.getCamerasSuccess>
) => {...}
```

I do have a similar problem in redux-observable actions, always writing a `if (action.type !== ...) ...`-guard. Maybe that will be solved with TS 2.8 too.

**Comment by piotrwitek** [comment]
Closing as this was solved here: https://github.com/piotrwitek/typesafe-actions/issues/19
